### PR TITLE
Ruby group capstone - Catalog of my things

### PR DESCRIPTION
**Pull request by HafizMassamTabraizKhan**
In this PR we did the following:

**LOGIC**

- Each class from the [UML class diagram](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/images/catalog_of_my_things.png) should be created in a separate .rb file.
- All associations (1-to-many relationships) and aggregations (parent-child relationships) between classes should be implemented.
- All properties visible in the diagram should be defined and set up in the constructor method. Exception: properties for the 1-to-many relationships should NOT be set in the constructor method. Instead, they should have a custom setter method created.
- All methods visible in the diagram should be implemented:
  - add_item method (in all classes that have that method)
    - should take an instance of the Item class as an input.
    - should add the input item to the collection of items.
    - should add self as a property of the item object (by using the correct setter from the item object).
  - can_be_archived?() in the Item class
    - should return true if published_date is older than 10 years.
    - otherwise, it should return false.
  - move_to_archive() in the Item class
    - should reuse can_be_archived?() method.
    - should change the archived property to true if the result of the can_be_archived?() method is true.
    - should do nothing if the result of the can_be_archived?() method is false.
  - can_be_archived?() in the Book class
    - should override the method from the parent class.
    - should return true if the parent's method returns true OR if cover_state equals to "bad".
    - otherwise, it should return false.
  - can_be_archived?() in the MusicAlbum class
    - should override the method from the parent class.
    - should return true if the parent's method returns true AND if on_spotify equals true.
    - otherwise, it should return false.
  - can_be_archived?() in the Movie class
    - should override the method from the parent class.
    - should return true if the parent's method returns true OR if silent equals true.
    - otherwise, it should return false.
  - can_be_archived?() in the Game class
    - should override the method from the parent class.
    - should return true if the parent's method returns true AND if last_played_at is older than 2 years.
    - otherwise, it should return false.
- Add unit tests for all implemented methods.

**User interface**

- Create a main.rb file that will serve as your console app entry point.
- Your console app, at the start, should:
  - Present the user with a list of options to perform.
  - Let users choose an option.
  - If needed, ask for parameters for the option.
  - Have a way to quit the app.
- The following options should be available:
  - List all books
  - List all music albums
  - List all movies
  - List of games
  - List all genres (e.g 'Comedy', 'Thriller')
  - List all labels (e.g. 'Gift', 'New')
  - List all authors (e.g. 'Stephen King')
  - List all sources (e.g. 'From a friend', 'Online shop')
  - Add a book
  - Add a music album
  - Add a movie
  - Add a game
- All data should be preserved by saving collections in .json files.

**Database**

- Create a schema.sql file with tables that will be analogical to the structure of the classes in our app. As we cannot implement inheritance in the database tables - add all properties and associations from the parent Item class as table columns to all tables based on the child classes. Thanks to that it can be used to store data in the future.

### CODE REVIEW

File: spec/label_spec.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+  describe '#add_item' do
+    it 'adds an item to the items array' do
+      label = Label.new(label_title, label_color)
+      item = Item.new(genre: 'Sci-Fi', author: 'Author', publish_date: '2000-01-01')
+
+      label.add_item(item)
+
+      expect(label.instance_variable_get(:@items)).to include(item)
+    end
```
Comment (codepantha) [inline spec/label_spec.rb]: - [x] All your tests in the `label_spec.rb` file are failing as a result of supplying the wrong number of arguments when initializing the `Label` class. Please try to fix this issue 👍 

image

File: schema.sql [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+CREATE TABLE books (
+  id INT GENERATED ALWAYS AS IDENTITY,
+  publish_date DATE,
+  label_id INT REFERENCES labels(id),
+  can_be_archived BOOLEAN,
+  cover_state VARCHAR(250),
+  publisher VARCHAR(250),
+  PRIMARY KEY(id)
+);
```
Comment (codepantha) [inline schema.sql]: - [x] Please add the `genre_id` and `author_id` columns to the books table as they should correspond to the properties from the UML diagram

File: schema.sql [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+CREATE TABLE music_album
+(
+  id INT PRIMARY KEY NOT NULL,
+  on_spotify BOOLEAN,
+  publish_date DATE,
+  archived BOOLEAN,
+  genre_id INT REFERENCES genre(id)
+);
```
Comment (codepantha) [inline schema.sql]: - [x] Please add the `label_id` and `author_id` columns to the `music_album` table as they should correspond to the properties from the UML diagram

File: schema.sql [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+CREATE TABLE games (
+  id SERIAL PRIMARY KEY,
+  publish_date DATE,
+  archived BOOLEAN,
+  multiplayer BOOLEAN,
+  played_at_date DATE,
+  author_id INT NOT NULL,
+  FOREIGN KEY (author_id) REFERENCES authors(id)
+);
```
Comment (codepantha) [inline schema.sql]: - [x] Please add the `genre_id` and `label_id` columns to the games table as they should correspond to the properties from the UML diagram

**Review by codepantha — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED ♻️ 

## Hi team 👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights 🎯 
✔️ Descriptive PR
✔️ App working as expected 📌 
✔️ All parent-child relationships are implemented 📌 
✔️ Good commit history
✔️ No linter errors

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- n/a

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by KayLemba — APPROVED** [review]
[[# STATUS: APPROVED :green_circle: :clap: 

Hello @TEAM    , 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:  . 
Congratulations! 🎉  Keep up the good work 🥂 

HIGHLIGHTS:

- [x]  Your repo is professional  😎 
- [x] Descriptive P.R 💯 
- [x] Database setup correctly
- [x] Professional Readme
- [x] Tests are passing 💯 
- [x] Professional Video Presentation.



 

 










 








Now you can merge this branch and move on :1st_place_medal: 
Happy Coding cheers 🥇 💯 



well-1

Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.



Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._]]

**Pull request merged**